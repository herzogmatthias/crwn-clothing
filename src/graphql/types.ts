import { ICartItem } from "../redux/cart/ICartItem";

export interface IHeaderType {
  cartHidden: boolean;
}

export interface ICacheState {
  cartHidden: boolean;
  cartItems: ICartItem[];
  itemCount: number;
}
