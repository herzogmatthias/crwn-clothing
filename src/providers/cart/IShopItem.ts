export interface IShopItem {
  id: number;
  name: string;
  imageUrl: string;
  price: number;
}
