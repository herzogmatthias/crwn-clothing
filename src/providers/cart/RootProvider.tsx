import * as React from "react";
import CartProvider from "./cart.provider";

export interface IRootProviderProps {
  children: React.ReactNode;
}

export default function RootProvider({ children }: IRootProviderProps) {
  return (
    <div>
      <CartProvider>{children}</CartProvider>
    </div>
  );
}
