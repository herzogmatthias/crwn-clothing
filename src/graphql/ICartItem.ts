import { IShopItem } from "./IShopItem";

export interface ICartItem extends IShopItem {
  quantity: number;
}
