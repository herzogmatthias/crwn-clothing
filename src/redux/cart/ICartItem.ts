import { IShopItem } from "../shop/IShopItem";

export interface ICartItem extends IShopItem {
  quantity: number;
}
