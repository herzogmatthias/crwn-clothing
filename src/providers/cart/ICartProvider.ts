import { ICartItem } from "../../redux/cart/ICartItem";
import { IShopItem } from "../../redux/shop/IShopItem";

export interface ICartProvider {
  hidden: boolean;
  total: number;
  toggleHidden(): void;
  cartItems: ICartItem[];
  addItem(item: IShopItem): void;
  removeItem(item: ICartItem): void;
  clearItemFromCart(item: ICartItem): void;
  cartItemsCount: number;
}
