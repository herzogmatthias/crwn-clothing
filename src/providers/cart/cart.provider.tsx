import React, { createContext, useEffect } from "react";
import { ICartProvider } from "./ICartProvider";
import { ICartItem } from "../../redux/cart/ICartItem";
import {
  addItemToCart,
  removeItemFromCart,
  filterItemFromCart,
  getCartItemsCount,
  getTotalPrice,
} from "../../redux/cart/cart.utils";
import { IShopItem } from "../../redux/shop/IShopItem";

export const CartContext = createContext<ICartProvider>({
  hidden: true,
  total: 0,
  cartItems: [],
  cartItemsCount: 0,
  toggleHidden: () => {},
  addItem: () => {},
  removeItem: () => {},
  clearItemFromCart: () => {},
});

interface ICartProviderProps {
  children: React.ReactNode;
}

const CartProvider = ({ children }: ICartProviderProps) => {
  const [hidden, setHidden] = React.useState(true);

  const [cartItems, setCartItems] = React.useState<ICartItem[]>([]);
  const [cartItemsCount, setCartItemsCount] = React.useState(0);

  const [total, setTotal] = React.useState(0);

  const toggleHidden = () => {
    setHidden(!hidden);
  };

  const addItem = (item: IShopItem) =>
    setCartItems(addItemToCart(cartItems, item));

  const removeItem = (item: ICartItem) =>
    setCartItems(removeItemFromCart(cartItems, item));
  const clearItemFromCart = (item: ICartItem) =>
    setCartItems(filterItemFromCart(cartItems, item));

  useEffect(() => {
    setCartItemsCount(getCartItemsCount(cartItems));
    setTotal(getTotalPrice(cartItems));
  }, [cartItems]);

  return (
    <CartContext.Provider
      value={{
        total,
        addItem,
        hidden,
        toggleHidden,
        cartItems,
        cartItemsCount,
        removeItem,
        clearItemFromCart,
      }}
    >
      {children}
    </CartContext.Provider>
  );
};

export default CartProvider;
