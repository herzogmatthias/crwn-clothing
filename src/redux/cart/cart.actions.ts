import { createAction } from "@reduxjs/toolkit";
import { TOGGLE_CART_HIDDEN, ADD_ITEM } from "./cart.types";
import { withPayloadType } from "../genericActionPayloadType";
import { IShopItem } from "../../pages/shop/IShopItem";
import { ICartItem } from "./ICartItem";

export const toggleCartHidden = createAction(TOGGLE_CART_HIDDEN);
export const addItem = createAction(ADD_ITEM, withPayloadType<IShopItem>());
