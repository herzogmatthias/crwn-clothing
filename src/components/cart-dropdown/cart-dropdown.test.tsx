import React from "react";
import { ReactWrapper, mount, shallow, ShallowWrapper } from "enzyme";
import configureStore, { MockStoreEnhanced } from "redux-mock-store";
import { ActionCreatorWithoutPayload } from "@reduxjs/toolkit";
import { TOGGLE_CART_HIDDEN } from "../../redux/cart/cart.types";
import {
  default as ConnectedCartDropdown,
  CartDropdown,
} from "./cart-dropdown.component";
import { Provider } from "react-redux";
import {
  CustomButtonContainer,
  EmptyMessageContainer,
} from "./cart-dropdown.styles";
import { BrowserRouter } from "react-router-dom";
import { ICartItem } from "../../redux/cart/ICartItem";
import CartItem from "../cart-item/cart-item.component";

describe("CONNECTED CARTDROPDOWN TESTS", () => {
  let wrapper: ReactWrapper;

  const cartItems: ICartItem[] = [
    { id: 1, imageUrl: "asdfasdf", name: "hats", price: 23, quantity: 2 },
    { id: 2, imageUrl: "asdfasdf", name: "hats", price: 23, quantity: 2 },
  ];

  let mockStore = configureStore([]);
  let store: MockStoreEnhanced<unknown, {}>;

  beforeEach(() => {
    const mockState = {
      cart: {
        cartItems: cartItems,
      },
    };

    store = mockStore(mockState);

    wrapper = mount(
      <BrowserRouter>
        <Provider store={store}>
          <ConnectedCartDropdown />
        </Provider>
      </BrowserRouter>
    );
  });
  test("render CartDropdown Component", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("check Props", () => {
    expect(wrapper.find(CartDropdown).prop("cartItems")).toEqual(cartItems);
  });
});

describe("NOT CONNECTED CARTDROPDOWN TESTS", () => {
  let wrapper: ShallowWrapper;
  let mockAction: ActionCreatorWithoutPayload<"TOGGLE_CART_HIDDEN">;
  let mockHistory: { push: Function };
  const push = jest.fn<any, [string]>((url) => true);
  const cartItems: ICartItem[] = [
    { id: 1, imageUrl: "asdfasdf", name: "hats", price: 23, quantity: 2 },
    { id: 2, imageUrl: "asdfasdf", name: "hats", price: 23, quantity: 2 },
  ];

  beforeEach(() => {
    const dispatch = jest.fn();
    mockAction = jest.fn(() => {
      dispatch({ type: TOGGLE_CART_HIDDEN, payload: undefined });
    }) as any;

    mockHistory = {
      push: push,
    };
    const mockProps = {
      history: mockHistory,
      toggleCartHidden: mockAction,
      cartItems: cartItems,
    } as any;
    wrapper = shallow(<CartDropdown {...mockProps}></CartDropdown>);
  });
  test("call history.push and dispatch action when button is clicked", () => {
    wrapper.find(CustomButtonContainer).simulate("click");
    expect(mockHistory.push).toHaveBeenCalledWith("/checkout");
    expect(mockAction).toHaveBeenCalled();
  });
  test("render an equal number of CartItem components as the cartItems prop", () => {
    expect(wrapper.find(CartItem).length).toEqual(cartItems.length);
  });
  test("render EmptyMessageContainer if cartItems is empty", () => {
    const mockProps = {
      history: mockHistory,
      toggleCartHidden: mockAction,
      cartItems: [],
    } as any;

    const newWrapper = shallow(<CartDropdown {...mockProps} />);
    expect(newWrapper.exists(EmptyMessageContainer)).toBe(true);
  });
});
