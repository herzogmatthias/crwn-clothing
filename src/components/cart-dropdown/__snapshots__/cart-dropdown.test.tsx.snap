// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CONNECTED CARTDROPDOWN TESTS render CartDropdown Component 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <withRouter(Connect(CartDropdown))>
        <Connect(CartDropdown)
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "createHref": [Function],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
        >
          <CartDropdown
            cartItems={
              Array [
                Object {
                  "id": 1,
                  "imageUrl": "asdfasdf",
                  "name": "hats",
                  "price": 23,
                  "quantity": 2,
                },
                Object {
                  "id": 2,
                  "imageUrl": "asdfasdf",
                  "name": "hats",
                  "price": 23,
                  "quantity": 2,
                },
              ]
            }
            history={
              Object {
                "action": "POP",
                "block": [Function],
                "createHref": [Function],
                "go": [Function],
                "goBack": [Function],
                "goForward": [Function],
                "length": 1,
                "listen": [Function],
                "location": Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
                "push": [Function],
                "replace": [Function],
              }
            }
            location={
              Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              }
            }
            match={
              Object {
                "isExact": true,
                "params": Object {},
                "path": "/",
                "url": "/",
              }
            }
            toggleCartHidden={[Function]}
          >
            <styled.div
              className="cart-dropdown"
            >
              <div
                className="sc-AxgMl feQnvF cart-dropdown"
              >
                <styled.div>
                  <div
                    className="sc-Axmtr fIyTqw"
                  >
                    <Memo(CartItem)
                      item={
                        Object {
                          "id": 1,
                          "imageUrl": "asdfasdf",
                          "name": "hats",
                          "price": 23,
                          "quantity": 2,
                        }
                      }
                      key="1"
                    >
                      <styled.div>
                        <div
                          className="sc-AxirZ dItcAc"
                        >
                          <styled.img
                            alt="item"
                            src="asdfasdf"
                          >
                            <img
                              alt="item"
                              className="sc-AxjAm eYwkFk"
                              src="asdfasdf"
                            />
                          </styled.img>
                          <styled.div>
                            <div
                              className="sc-AxiKw fKyxRv"
                            >
                              <styled.span>
                                <span
                                  className="sc-AxhCb dBPtdb"
                                >
                                  hats
                                </span>
                              </styled.span>
                              <span
                                className="price"
                              >
                                2
                                 x $
                                23
                              </span>
                            </div>
                          </styled.div>
                        </div>
                      </styled.div>
                    </Memo(CartItem)>
                    <Memo(CartItem)
                      item={
                        Object {
                          "id": 2,
                          "imageUrl": "asdfasdf",
                          "name": "hats",
                          "price": 23,
                          "quantity": 2,
                        }
                      }
                      key="2"
                    >
                      <styled.div>
                        <div
                          className="sc-AxirZ dItcAc"
                        >
                          <styled.img
                            alt="item"
                            src="asdfasdf"
                          >
                            <img
                              alt="item"
                              className="sc-AxjAm eYwkFk"
                              src="asdfasdf"
                            />
                          </styled.img>
                          <styled.div>
                            <div
                              className="sc-AxiKw fKyxRv"
                            >
                              <styled.span>
                                <span
                                  className="sc-AxhCb dBPtdb"
                                >
                                  hats
                                </span>
                              </styled.span>
                              <span
                                className="price"
                              >
                                2
                                 x $
                                23
                              </span>
                            </div>
                          </styled.div>
                        </div>
                      </styled.div>
                    </Memo(CartItem)>
                  </div>
                </styled.div>
                <Styled(CustomButton)
                  onClick={[Function]}
                >
                  <CustomButton
                    className="sc-AxheI isBHfp"
                    inverted={false}
                    isGoogleSignIn={false}
                    onClick={[Function]}
                  >
                    <styled.button
                      className="sc-AxheI isBHfp"
                      inverted={false}
                      isGoogleSignIn={false}
                      onClick={[Function]}
                    >
                      <button
                        className="sc-AxhUy hdAkhk sc-AxheI isBHfp"
                        onClick={[Function]}
                      >
                        GO TO CHECKOUT
                      </button>
                    </styled.button>
                  </CustomButton>
                </Styled(CustomButton)>
              </div>
            </styled.div>
          </CartDropdown>
        </Connect(CartDropdown)>
      </withRouter(Connect(CartDropdown))>
    </Provider>
  </Router>
</BrowserRouter>
`;
